### ISSUE TITLE
support for pattern-inside a control flow

### ISSUE THREAD

**Issue by gordonwoodhull** [semgrep/semgrep#3801]
**Describe the bug**
We'd like to use semgrep to do the analysis of control flow that we can (in theory) do with SmPL. I'm not clear if semgrep patterns follow control flow or if it is only syntactic/AST matching.

Say we have some `Foo` provider and we want to match the region from the construction of the `Foo` to the call of `Foo.bar()`:

```yaml
  pattern: |
    $FOO = $PROVIDER.foo();
    ...
    $FOO.bar();
```

This matches the "no blocks" case:
```java
class C {
    public void f() {
        Foo foo = provider.foo();
        something.qux();
        foo.bar();
    }
}
```

and the "ends inside block" case:
```java
class C {
    public void f() {
        Foo foo = provider.foo();
        something.qux();
        if(true) {
            foo.bar();
        }
    }
}
```

but not the "starts in block" case:

```java
class C {
    public void f() {
        Foo foo;
        if(true) {
            foo = provider.foo();
        }
        something.qux();
        foo.bar();
    }
}
```

I see the same behavior for try-blocks, our actual use case. I'm using `if` for simplicity here.

**To Reproduce**
https://semgrep.dev/s/gordonwoodhull:ellipsis-control-flow

**What is the priority of the bug to you?**

P1. We have had great success applying vulnerability remediation fixes to code using semgrep, but we want more control over which patches are selected and where they are applied. I am hoping to use `pattern-inside` to specify a control-flow pattern which should contain the statement to be patched.

**Comment by ievans** [comment]
@gordonwoodhull this looks like a use-case for the new taint mode, which was recently promoted from experimental (but is not yet GA). How does https://semgrep.dev/s/ievans:java-taint-cfg look to you :D 

Docs: https://semgrep.dev/docs/writing-rules/data-flow/

**Comment by gordonwoodhull** [comment]
Thanks @ievans! That's a neat feature, and I guess it confirms my guess that ordinary patterns and ellipsis are syntactic and do not follow control flow.

Unfortunately I am looking for something with deeper integration with autofix. Specifically in our use case we would like to
- modify (or prepend statements) to the source statement (not just the sink)
- use bound metavariables in the `fix`

That's why I was hoping to use `pattern-inside`. I suppose I will fall back to writing external tools that use line numbers reported by semgrep.

We don't need interprocedural analysis. It looks like ellipsis matches intraprocedural control flow if it is outer-to-inner scope, but I can see why inner-to-outer might be be "out of scope".

**Comment by ievans** [comment]
@gordonwoodhull thanks for sharing. It sounds like the refactor you want to perform might be quite complex, so using Semgrep as a tool in a pipeline makes sense to me. 

At the same time, I'm curious if you have an imagined syntax that would allow Semgrep to perform the refactor. It sounds like in the search (as opposed to taint mode) you want:

```
pattern:
    $FOO = $PROVIDER.foo();
    ...
    $FOO.bar();
```

```
fix:
    some_new_stmt;
    $FOO = $PROVIDER.foo();
    ...
    $FOO.bar();
```

Is that right? How would you imagine using the `taint` mode to express the refactor?

(and cc @IagoAbal who I bet would have some thoughts)

**Comment by gordonwoodhull** [comment]
Thanks @ievans.

Yes, I want something like that, but it is impossible for `...` to appear in a fix (what would it expand to?), so I was trying to use `pattern-inside` to specify the control flow, and `pattern` to express the anchor line for the insertion:

```yaml
  patterns:
    - pattern-inside: |
        $FOO = $PROVIDER.foo();
        ...
        $FOO.bar();
    - pattern: |
        $FOO = $PROVIDER.foo();
  fix: |
    // some statements using $PROVIDER and $FOO
    $FOO = $PROVIDER.foo();
```

This again has the desired behavior with the "no blocks" and "ends in block" cases, but does not match the "starts in block" case.

https://semgrep.dev/s/gordonwoodhull:ellipsis-control-flow-pattern-inside

#### Bigger picture
1. There may be more "context" lines which must be in the right order in the control flow for the pattern to match, but there is just one line which needs to be changed or used as an anchor for inserting statements.
2. A reasonable dialect of [SmPL](https://who.paris.inria.fr/Julia.Lawall/ercim.pdf) can be implemented on top of Semgrep, generating the above rule from the succinct syntax

```patch
+    // some statements using $PROVIDER and $FOO
     $FOO = $PROVIDER.foo();
     ...
     $FOO.bar();
```
3. I am willing and may be competent to contribute to Semgrep to make this functionality happen in some form or another.

**Comment by IagoAbal** [comment]
cc @aryx who may have some thoughts too

Semgrep search mode works by traversing the AST and trying to match patterns against nodes along the way. Indeed it does not follow the CFG as Coccinelle does, hence the behavior that you observed. We could make Semgrep work the same way as Coccinelle, although it's probably quite some work. Our approach so far has been to keep matching simple, but complement Semgrep with some flow analyses.

Now, in order to solve your problem, perhaps we could allow something like this:

    patterns:
      - pattern-inside-taint:
          pattern-sources:
          - pattern: $PROVIDER.foo();
          pattern-sinks:
          - pattern: $FOO.bar();
      - pattern: |
          $FOO = $PROVIDER.foo();
    fix: |
      // some statements using $PROVIDER and $FOO
      $FOO = $PROVIDER.foo();

I think this is doable in Semgrep right now, since patterns return ranges instead of ASTs, so we should be able to use a taint-rule inside a search-rule in order to delimit a region of text, even if it doesn't correspond to a node in the AST.

> There may be more "context" lines which must be in the right order in the control flow for the pattern to match, but there is just one line which needs to be changed or used as an anchor for inserting statements.

Could you provide some examples here @gordonwoodhull? This is something that we don't support in taint-mode yet, but it was in my to-do list to write a proposal for extension. It would be good to understand what you need here  so we try to cover it as well.

**Comment by gordonwoodhull** [comment]
Thanks @IagoAbal. This is promising.

I think it would solve many or most of our use cases, as long as we don't need more than two matches in the control flow. (The equivalent of multiple `...` in SmPL.) <strike>I will check on that.</strike> **EDIT:** as far as I can tell, we only need a "line of interest" and a line before/after it that needs to be changed, so yes this works.

One detail: it's important to have

```yaml
      pattern-sources:
      - pattern: $FOO = $PROVIDER.foo();     # not just $PROVIDER.foo()
      pattern-sinks:
      - pattern: $FOO.bar();
```

I.e. we want to match a specific flow where some `$FOO` was initialized and then later `.bar()` was called on that same `$FOO`. I just tested this and indeed it looks like metavariables are shared between `pattern-sources` and `pattern-sinks` in taint mode, so this works too. ([Here is the rule](https://semgrep.dev/s/gordonwoodhull:metavariables-in-taint-mode-2), which correctly fails if  `$FOO` is not the same.)

I take it taint mode is interprocedural at the file level, so we might need to narrow the matches (since a range that spans multiple functions would not make sense).

**Comment by gordonwoodhull** [comment]
Thanks @IagoAbal for the suggestion of how to implement this functionality. I implemented it successfully!

I will offer an experience report now, and if my project continues next year I will clean up my changes, write tests, and contribute a PR.

I found the Semgrep code surprisingly easy to read given how powerful it is. Well commented and just the needed abstractions. Kudos!

Okay, here are the difficulties I encountered implementing this, for discussion later if I manage to contribute, otherwise for anyone interested in this feature.

* The obvious way to implement `pattern-inside-taint` is to create a new `xpattern_kind` which calls the existing `check_taint`. The problem is that `Tainting_generic` calls `matches_of_formula` for the source/sink/sanitizer patterns. Mutual recursion between modules.
  Similarly, we end up with a big new circular type dependency in `Rule.ml` because a taint spec contains patterns, and `pattern-inside-taint` is a kind of pattern. It doesn't make sense for `pattern-inside-taint` to include `pattern-inside-taint` but that's how the types are set up today.
  I worked around these problems in the obvious abominable way, but I will refactor if I contribute. My guess is that the taint sub-patterns should be handled in the ordinary xpatterns phase, and then `pattern-inside-taint` should run, and then the boolean operations should run. I haven't thought about the circular types - those seem hard to avoid.

* I was hung up for a while not having the source for each sink, but #4703 arrived while I was working on this. Thanks @mmcqd! Using the `PM.Set.t` I was able to construct the full ranges using the source info.

* Minor trouble but worth mentioning: In addition to changing the `range_loc` of each pattern match to cover source->sink, the `pattern_id` also has to be changed to that of the parent `pattern-inside-taint`. So I changed the input to `(taint_rules : (pattern_id option * R.t * Rule.taint_spec) list)` and changed `found_tainted_sink` to make these transformations. 
  Probably `found_tainted_sink` should be exposed as a parameter instead.

For now I am only creating a proof-of-concept of using Semgrep for remediating vulnerabilities in Java, but if the company moves forward with this, we'll want to contribute a bunch of features and fixes.

Next I'm going to look at #3577 multiple autofix on a line. (I suspect it has the same root cause as #3388 multiple multiline autofixes.)

I also have another feature request called "patterns as anchors" which I will file for discussion soon.

Thanks for an amazing tool!

**Comment by IagoAbal** [comment]
Wow, very nice @gordonwoodhull ! I am looking forward to seeing that PR.

I think that if we add something like `pattern-inside-taint` then those circular dependencies are natural and we'll deal with them. I think we had a bit of that when we added `metavariable-pattern` too. Mutually recursive types are not a problem. But we may want to merge `Tainting_generic` with `Match_rules`.

Let's think about it more carefully when you open that PR. (I hope you do!)

**Comment by gordonwoodhull** [comment]
:wave:

This is high-priority for us, we have implemented it, and plan to contribute. Since the project is continuing this year, we will get approved to contribute soon!

**Comment by aryx** [comment]
woohoo, great!

**Comment by gordonwoodhull** [comment]
I found time over the holidays to catch up with what's changed and look at how to contribute this.

It turns out most of the work has already been done in #5993 taint patterns, and more nicely than my implementation. But that PR stalled.

I've submitted a discussion/demo PR #6764 showing the additional change needed for this feature, along with tests. It would need a flag or an option because the range spans from source to sink instead of just including the sink.

I'd be glad to put in the effort to get this into Semgrep, if there is interest.
